routing/payment_lifecycle.go

```diff
@@ -26,166 +28,445 @@ func (e errNoRoute) Error() string {
 }
 
+// paymentFailure represents a terminal failure that can happen during the
+// payment lifecycle. We use it to record such an encountered error while we
+// wait for the remaining shards to finish.
+type paymentFailure struct {
+	failureCode channeldb.FailureReason
+	err         error
+}
+
 // paymentLifecycle holds all information about the current state of a payment
 // needed to resume if from any point.
 type paymentLifecycle struct {
-	router         *ChannelRouter
-	payment        *LightningPayment
-	paySession     PaymentSession
-	timeoutChan    <-chan time.Time
-	currentHeight  int32
-	finalCLTVDelta uint16
-	attempt        *channeldb.PaymentAttemptInfo
-	circuit        *sphinx.Circuit
-	lastError      error
+	router           *ChannelRouter
+	totalAmount      lnwire.MilliSatoshi
+	paymentHash      lntypes.Hash
+	paySession       PaymentSession
+	timeoutChan      <-chan time.Time
+	currentHeight    int32
+	existingAttempts []channeldb.HTLCAttemptInfo
+	lastError        error
+
+	// quitCollectShard is closed to signal the collectShard sub goroutines
+	// of the payment lifecycle to stop.
+	quitCollectShard chan struct{}
+	wgCollectShard   sync.WaitGroup
 }
 
 // resumePayment resumes the paymentLifecycle from the current state.
 func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
-	// We'll continue until either our payment succeeds, or we encounter a
-	// critical error during path finding.
+	// When the payment lifecycle loop exits, we make sure to signal any
+	// sub goroutine to exit, then wait for them to return.
+	defer func() {
+		close(p.quitCollectShard)
+		p.wgCollectShard.Wait()
+	}()
+
+	// Cancel the payment session when we exit to signal underlying
+	// goroutines can exit.
+	defer p.paySession.Cancel()
+
+	// Each time we send a new payment shard, we'll spin up a goroutine
+	// that will collect the result. Either a payment result will be
+	// returned, or a critical error signaling that we should immediately
+	// exit.
+	shardResults := make(chan *shardResult)
+	criticalErr := make(chan error, 1)
+	shards := &PaymentShards{
+		PaymentHash: p.paymentHash,
+		Control:     p.router.cfg.Control,
+	}
+
+	// If we had any existing attempts outstanding, add them to our set of
+	// outstanding shards and resume their goroutines such that the final
+	// result will be given to the lifecycle loop below.
+	for _, a := range p.existingAttempts {
+		err := shards.AddShard(
+			&a,
+			make(chan *RouteResult, 1), // Result will be ignored.
+		)
+		if err != nil {
+			return [32]byte{}, nil, err
+		}
+
+		p.wgCollectShard.Add(1)
+		go p.collectShard(&a, shardResults, criticalErr)
+	}
+
+	// We'll continue sending shards until either our payment succeeds, or
+	// we encounter an error during path finding, or another terminal
+	// error.
+	var (
+		success         = false
+		routeFailure    *paymentFailure
+		terminalFailure *paymentFailure
+	)
 	for {
+		// Check if we got any terminal failure.
```

Comment:
Because the payment loop goroutine remains active and uses up resources, it always needs to end. This necessitates a somewhat artificial timeout on the sendtoroute-paymentsession. An idea could be to not have an active payment loop, but let the per-htlc goroutines control the payment process. Whenever an htlc comes back, the current state of the payment is evaluated. The amount and fee budget remaining is calculated directly from the payment in the database and based on that zero, one or multiple additional htlcs are launched.

This would remove the need for the payment session timeout and possibly also simplify the code. Separate tracking of shards in-flight, amount remaining etc is no longer required, reducing the potential for bugs and desynchronization.

If we want this, this is the time to do it. We won't overhaul the core of payment sending again anytime soon I expect, given the size of the change set in this pr.

Reply:
Done.